app/lib/search/backend.dart

```diff
@@ -76,3 +91,102 @@ class SearchBackend {
 }
 
 String _toUrl(String package) => 'https://pub.dartlang.org/packages/$package';
+
+class SnapshotStorage {
+  final Storage storage;
+  final Bucket bucket;
+  final GZipCodec _gzip = new GZipCodec();
+
+  SnapshotStorage(this.storage, this.bucket);
+
+  Future<SearchSnapshot> fetch() async {
+    final List<BucketEntry> list = await bucket.list().toList();
+    final List<String> names = list
+        .where((entry) => entry.isObject)
+        .map((entry) => entry.name)
+        .toList();
+    if (names.isEmpty) return null;
+    // Select the first entry, because there is chance that later entries were
+    // aborted for some reason, and the most probable correct entry is the first
+    // one.
+    names.sort();
+    final String selected = names.first;
+    // reading
+    final Stream<List<int>> storageStream = bucket.read(selected);
+    final Stream<List<int>> unzippedStream = _gzip.decoder.bind(storageStream);
+    final String json = await UTF8.decoder.bind(unzippedStream).join();
+    return new SearchSnapshot.fromJson(JSON.decode(json));
+  }
+
+  Future store(SearchSnapshot snapshot) async {
+    // garbage-collect old entries after upload is successful
+    final List<BucketEntry> list = await bucket.list().toList();
+    final List<String> toDelete = list
+        .where((entry) => entry.isObject)
+        .map((entry) => entry.name)
+        .toList();
+
+    // data buffer to write
+    final List<int> buffer =
+        _gzip.encode(UTF8.encode(JSON.encode(snapshot.toJson())));
+
+    // generate name from current timestamp
+    final String ts =
+        new DateTime.now().toUtc().toIso8601String().replaceAll(':', '-');
+    final currentName = 'snapshot-$ts.json.gz';
+
+    // upload
+    final StreamSink<List<int>> sink = bucket.write(currentName);
+    sink.add(buffer);
+    await sink.close();
+
+    // upload successful, garbage-collect entries
+    for (String name in toDelete) {
+      try {
+        await bucket.delete(name);
+      } catch (e, st) {
+        _logger.warning('Snapshot delete failed: $name', e, st);
+      }
+    }
+  }
+}
+
+class SearchSnapshot {
+  DateTime updated;
```

Comment:
it could, but it would make tracking the last updated timestamp much more complex and awkward. so, no, can't. :)